File: ts/src/test/base/test.safeTicker.ts



import assert from 'assert';
import Precise from '../../base/Precise.js';
import ccxt from '../../../ccxt.js';



function preciseEqualStr (exchange:any, result: any, key: string, expected: string) {
    return Precise.stringEq (exchange.safeString (result, key), expected);
}

// asserts a flat [ field, expected, field, expected, ... ] list against a safeTicker result
function preciseEqualFields (exchange: any, result: any, pairs: string[]) {
    assert (pairs.length % 2 === 0, 'the field/expected list must come in pairs');
    for (let i = 0; i < pairs.length; i++) {
        if (i % 2 === 0) {
            const key = pairs[i];
            const expected = pairs[i + 1];
            assert (preciseEqualStr (exchange, result, key, expected), 'safeTicker ' + key + ' must be ' + expected);
        }
    }
}

function testSafeTicker () {

    const exchange = new ccxt.Exchange ({
        'id': 'sampleexchange',
    });

    // CASE 1 - by open
    const ticker1 = {
        'open': 5.0,
        'change': 1.0,
    };
    const result1 = exchange.safeTicker (ticker1);
    preciseEqualFields (exchange, result1, [ 'percentage', '20.0', 'average', '5.5', 'close', '6.0', 'last', '6.0' ]);

    // CASE 2 - by open
    const ticker2 = {
        'open': 5.0,
        'percentage': 20.0,
    };
    const result2 = exchange.safeTicker (ticker2);
    preciseEqualFields (exchange, result2, [ 'change', '1.0', 'average', '5.5', 'close', '6.0', 'last', '6.0' ]);

    // CASE 3 - by close
    const ticker3 = {
        'close': 6.0,
        'change': 1.0,
    };
    const result3 = exchange.safeTicker (ticker3);
    preciseEqualFields (exchange, result3, [ 'open', '5.0', 'percentage', '20.0', 'average', '5.5', 'last', '6.0' ]);

    // CASE 4 - by close
    const ticker4 = {
        'close': 6.0,
        'percentage': 20.0,
    };
    const result4 = exchange.safeTicker (ticker4);
    preciseEqualFields (exchange, result4, [ 'open', '5.0', 'change', '1.0', 'average', '5.5', 'last', '6.0' ]);

    // CASE 5 - by average
    const ticker5 = {
        'average': 5.5,
        'percentage': 20.0,
    };
    const result5 = exchange.safeTicker (ticker5);
    preciseEqualFields (exchange, result5, [ 'open', '5.0', 'change', '1.0', 'close', '6.0', 'last', '6.0' ]);

    // CASE 6
    const ticker6 = {
        'average': 5.5,
        'change': 1.0,
    };
    const result6 = exchange.safeTicker (ticker6);
    preciseEqualFields (exchange, result6, [ 'open', '5.0', 'percentage', '20.0', 'close', '6.0', 'last', '6.0' ]);

    // CASE 7 - by open and close
    const ticker7 = {
        'open': 5.0,
        'close': 6.0,
    };
    const result7 = exchange.safeTicker (ticker7);
    preciseEqualFields (exchange, result7, [ 'change', '1.0', 'percentage', '20.0', 'average', '5.5', 'last', '6.0' ]);

    // CASE 8 - full ticker
    const ticker8 = {
        'open': 5.0,
        'close': 6.0,
        'last': 6.0,
        'high': 6.5,
        'low': 4.5,
        'average': 5.5,
        'bid': 5.9,
        'bidVolume': 100,
        'ask': 6.1,
        'askVolume': 200,
        'change': 1.0,
        'percentage': 20.0,
        'vwap': 5.75,
        'baseVolume': 1000,
        'quoteVolume': 5750,
        'previousClose': 4.9,
        'indexPrice': 5.8,
        'markPrice': 5.9,
        'info': {}
    };
    const result8 = exchange.safeTicker (ticker8);
    preciseEqualFields (exchange, result8, [ 'open', '5.0', 'high', '6.5', 'low', '4.5', 'close', '6.0', 'last', '6.0', 'change', '1.0', 'percentage', '20.0', 'average', '5.5', 'bid', '5.9', 'bidVolume', '100.0', 'ask', '6.1', 'askVolume', '200.0', 'vwap', '5.75', 'baseVolume', '1000.0', 'quoteVolume', '5750.0', 'previousClose', '4.9', 'indexPrice', '5.8', 'markPrice', '5.9' ]);
    assert (result8['info'] !== undefined);

    // CASE 9 - flat day, a legitimate zero change must be preserved, see https://github.com/ccxt/ccxt/issues/25971
    const ticker9 = {
        'open': 6.0,
        'close': 6.0,
        'last': 6.0,
        'change': 0.0,
        'percentage': 0.0,
    };
    const result9 = exchange.safeTicker (ticker9);
    preciseEqualFields (exchange, result9, [ 'change', '0', 'percentage', '0', 'open', '6.0', 'last', '6.0' ]);

    // CASE 10 - by open and average, the pair that derives close from average
    const ticker10 = {
        'open': 5.0,
        'average': 5.5,
    };
    const result10 = exchange.safeTicker (ticker10);
    preciseEqualFields (exchange, result10, [ 'close', '6.0', 'last', '6.0', 'average', '5.5' ]);
    // the supplied average survives untouched, and this path leaves change and percentage underived
    assert (result10['change'] === undefined);
    assert (result10['percentage'] === undefined);
}

export default testSafeTicker;

File: julia/Ccxt/test/base/test.safeTicker.jl
using Test
using Ccxt
function preciseEqualStr(exchange, result, key, expected)

    return stringEq(safeString(exchange, result, key), expected)
end


function testSafeTicker()

    exchange = get(ccxt, Symbol("Exchange"), nothing)(Dict{Symbol, Any}(
        Symbol("id") => "sampleexchange"
    ));
    ticker1 = Dict{Symbol, Any}(
        Symbol("open") => 5,
        Symbol("change") => 1
    );
    result1 = safeTicker(exchange, ticker1);
    @test functions.ccxtruthy(preciseEqualStr(exchange, result1, "percentage", "20.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result1, "average", "5.5"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result1, "close", "6.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result1, "last", "6.0"));
    ticker2 = Dict{Symbol, Any}(
        Symbol("open") => 5,
        Symbol("percentage") => 20
    );
    result2 = safeTicker(exchange, ticker2);
    @test functions.ccxtruthy(preciseEqualStr(exchange, result2, "change", "1.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result2, "average", "5.5"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result2, "close", "6.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result2, "last", "6.0"));
    ticker3 = Dict{Symbol, Any}(
        Symbol("close") => 6,
        Symbol("change") => 1
    );
    result3 = safeTicker(exchange, ticker3);
    @test functions.ccxtruthy(preciseEqualStr(exchange, result3, "open", "5.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result3, "percentage", "20.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result3, "average", "5.5"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result3, "last", "6.0"));
    ticker4 = Dict{Symbol, Any}(
        Symbol("close") => 6,
        Symbol("percentage") => 20
    );
    result4 = safeTicker(exchange, ticker4);
    @test functions.ccxtruthy(preciseEqualStr(exchange, result4, "open", "5.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result4, "change", "1.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result4, "average", "5.5"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result4, "last", "6.0"));
    ticker5 = Dict{Symbol, Any}(
        Symbol("average") => 5.5,
        Symbol("percentage") => 20
    );
    result5 = safeTicker(exchange, ticker5);
    @test functions.ccxtruthy(preciseEqualStr(exchange, result5, "open", "5.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result5, "change", "1.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result5, "close", "6.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result5, "last", "6.0"));
    ticker6 = Dict{Symbol, Any}(
        Symbol("average") => 5.5,
        Symbol("change") => 1
    );
    result6 = safeTicker(exchange, ticker6);
    @test functions.ccxtruthy(preciseEqualStr(exchange, result6, "open", "5.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result6, "percentage", "20.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result6, "close", "6.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result6, "last", "6.0"));
    ticker7 = Dict{Symbol, Any}(
        Symbol("open") => 5,
        Symbol("close") => 6
    );
    result7 = safeTicker(exchange, ticker7);
    @test functions.ccxtruthy(preciseEqualStr(exchange, result7, "change", "1.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result7, "percentage", "20.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result7, "average", "5.5"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result7, "last", "6.0"));
    ticker8 = Dict{Symbol, Any}(
        Symbol("open") => 5,
        Symbol("close") => 6,
        Symbol("last") => 6,
        Symbol("high") => 6.5,
        Symbol("low") => 4.5,
        Symbol("average") => 5.5,
        Symbol("bid") => 5.9,
        Symbol("bidVolume") => 100,
        Symbol("ask") => 6.1,
        Symbol("askVolume") => 200,
        Symbol("change") => 1,
        Symbol("percentage") => 20,
        Symbol("vwap") => 5.75,
        Symbol("baseVolume") => 1000,
        Symbol("quoteVolume") => 5750,
        Symbol("previousClose") => 4.9,
        Symbol("indexPrice") => 5.8,
        Symbol("markPrice") => 5.9,
        Symbol("info") => Dict{Symbol, Any}()
    );
    result8 = safeTicker(exchange, ticker8);
    @test functions.ccxtruthy(preciseEqualStr(exchange, result8, "open", "5.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result8, "high", "6.5"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result8, "low", "4.5"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result8, "close", "6.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result8, "last", "6.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result8, "change", "1.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result8, "percentage", "20.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result8, "average", "5.5"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result8, "bid", "5.9"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result8, "bidVolume", "100.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result8, "ask", "6.1"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result8, "askVolume", "200.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result8, "vwap", "5.75"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result8, "baseVolume", "1000.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result8, "quoteVolume", "5750.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result8, "previousClose", "4.9"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result8, "indexPrice", "5.8"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result8, "markPrice", "5.9"));
    @test get(result8, Symbol("info"), nothing) != nothing;
    ticker9 = Dict{Symbol, Any}(
        Symbol("open") => 6,
        Symbol("close") => 6,
        Symbol("last") => 6,
        Symbol("change") => 0,
        Symbol("percentage") => 0
    );
    result9 = safeTicker(exchange, ticker9);
    @test functions.ccxtruthy(preciseEqualStr(exchange, result9, "change", "0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result9, "percentage", "0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result9, "open", "6.0"));
    @test functions.ccxtruthy(preciseEqualStr(exchange, result9, "last", "6.0"));
end
